drivers/gpio/gpio_stm32.c

```diff
@@ -39,28 +43,55 @@ static void gpio_stm32_isr(int line, void *arg)
 /**
  * @brief Common gpio flags to custom flags
  */
-const int gpio_stm32_flags_to_conf(int flags, int *pincfg)
+static int gpio_stm32_flags_to_conf(int flags, int *pincfg)
 {
-	int direction = flags & GPIO_DIR_MASK;
-	int pud = flags & GPIO_PUD_MASK;
 
 	if (pincfg == NULL) {
 		return -EINVAL;
 	}
 
-	if (direction == GPIO_DIR_OUT) {
+	if ((flags & GPIO_OUTPUT) != 0) {
+		/* Output only or Output/Input */
+
 		*pincfg = STM32_PINCFG_MODE_OUTPUT;
-	} else {
-		/* pull-{up,down} maybe? */
+
+		if ((flags & GPIO_SINGLE_ENDED) != 0) {
+			if (flags & GPIO_LINE_OPEN_DRAIN) {
+				*pincfg |= STM32_PINCFG_OPEN_DRAIN;
+			} else  {
+				/* Output can't be open source */
+				return -ENOTSUP;
+			}
+		} else if ((flags & GPIO_PUSH_PULL) != 0) {
+			*pincfg |= STM32_PINCFG_PUSH_PULL;
+		}
+
+		if ((flags & GPIO_PULL_UP) != 0) {
+			*pincfg |= STM32_PINCFG_PULL_UP;
+		} else if ((flags & GPIO_PULL_DOWN) != 0) {
+			*pincfg |= STM32_PINCFG_PULL_DOWN;
+		}
+
+	} else if  ((flags & GPIO_INPUT) != 0) {
+		/* Input */
+
+		if ((flags & GPIO_LINE_OPEN_DRAIN) != 0) {
```

Comment:
We should implement this check directly in the API as an assert (guarding against a case when output is not enabled but `GPIO_LINE_OPEN_DRAIN` is set). Let's remove it from the driver.